common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -979,81 +1093,127 @@ private boolean ensureThreadStarted(int oldState) {
     }
 
     private void doStartThread() {
-        assert thread == null;
         executor.execute(new Runnable() {
             @Override
             public void run() {
+                processingLock.lock();
+                assert thread == null;
                 thread = Thread.currentThread();
                 if (interrupted) {
                     thread.interrupt();
+                    interrupted = false;
                 }
-
                 boolean success = false;
                 updateLastExecutionTime();
+                boolean suspend = false;
                 try {
-                    SingleThreadEventExecutor.this.run();
-                    success = true;
+                    for (;;) {
+                        SingleThreadEventExecutor.this.run();
+                        success = true;
+
+                        int currentState = state;
+                        if (canSuspend(currentState)) {
+                            if (!STATE_UPDATER.compareAndSet(SingleThreadEventExecutor.this,
+                                    ST_SUSPENDING, ST_SUSPENDED)) {
+                                // Try again as the CAS failed.
+                                continue;
+                            }
+
+                            if (!canSuspend(ST_SUSPENDED)) {
+                                // Seems like there was something added to the task queue again in the meantime,
+                                // let's try to update the state to started and try again to execute the run loop.
+                                STATE_UPDATER.compareAndSet(SingleThreadEventExecutor.this,
+                                        currentState, ST_STARTED);
+                                continue;
+                            }
+                            suspend = true;
+                        }
+                        break;
+                    }
                 } catch (Throwable t) {
                     logger.warn("Unexpected exception from an event executor: ", t);
                 } finally {
-                    for (;;) {
-                        int oldState = state;
-                        if (oldState >= ST_SHUTTING_DOWN || STATE_UPDATER.compareAndSet(
-                                SingleThreadEventExecutor.this, oldState, ST_SHUTTING_DOWN)) {
-                            break;
+                    boolean shutdown = false;
```

Comment:
It seems to me that this could be simplified as `final boolean shutdown = !suspended;`: can either `shutdown` or `suspended` be removed?

Reply:
its in a finally block... so I think in theory it could mess up things. Better safe than sorry.

Reply:
you were right... changed it